FEATURE: Split the User Options to hide profile and presence

counterleft opened the pull request:
This PR addresses [this issue](https://meta.discourse.org/t/split-profile-hiding-and-presence/184316) of wanting to set your "hide profile" setting independently from your "hide presence" setting.

The PR splits this user option into two, from `UserOption#hide_profile_and_presence` to `#hide_profile` and `#hide_presence`.

The code changes touch a lot of files (b/c localization is involved). I would recommend that reviewers read through each commit separately to get a better understanding the set of the changes. Overall, the commits are ordered in the following way:

1. Add the `UserOption#hide_profile` attribute and propagate that up the stack (UI, import tasks).
2. Add the `UserOption#hide_presence` attribute and propagate that up the stack (UI, import tasks, plugins).
3. Remove code references to the old option, `hide_profile_and_presence` and fill the values for `hide_profile` and `hide_presence` for existing users based on their existing `hide_profile_and_presence` value.
4. Remove the DB column, `user_options.hide_profile_and_presence` as a post-migration.

Each commit has more detail in their commit message.

Please note that only the English (`en`) localization has been made. Blank stub values have been created for all other languages. I would like guidance on how we want to adjust the translations for "Hide my profile" and "Hide my presence".

Since there is a destructive operation here (removing an existing DB column), I can also make a separate PR for that operation (if that is helpful).

The `chat` and `discourse_presence` have also been updated to use the new `hide_presence` option.

counterleft commented:
Am I expected to see a complete CI run for this PR? I would love if that were possible. I ran the specs that I touched and created but did not run the full suite; I had a number of failed specs in my local environment on a fresh `main` clone.

gschlager commented:
> Please note that only the English (`en`) localization has been made. Blank stub values have been created for all other languages. I would like guidance on how we want to adjust the translations for "Hide my profile" and "Hide my presence".

There's no need to touch non-English locale files. Discourse will fall back to English if strings are missing in another language.

counterleft commented:
> There's no need to touch non-English locale files. Discourse will fall back to English if strings are missing in another language.

Ah 👍, good to know. Unless I hear differently, I'll undo the addition of the stub L10N values, debug the failing tests, and resolve the merge conflict with `users_controller_spec`. Will most likely end up force-pushing to this PR.

counterleft commented:
Ok, updates:

1. I removed the blank stub values the L10N keys for `hide_profile` and `hide_presence` per feedback.
2. The broken tests were from the discourse-follow and discourse-whos-online plugins. I didn't account for those earlier; I didn't know there was a separate repo of other official plugins. It looks like both plugins rely on `hide_profile_and_presence` either through the model attribute or via SQL. So I decided to keep `hide_profile_and_presence` as an available attribute so that the plugins have time to update themselves to use `hide_profile` and/or `hide_presence` directly.  I added a `before_save` callback that ensures that setting a new value on `hide_profile_and_presence` will update `hide_profile` and `hide_presence` (and vice versa the other way).

I _just_ realized that have broken core tests in the HEAD of this PR. Will reply again when I have those fixed.

counterleft commented:
RE: the [test failures on this run](https://github.com/discourse/discourse/actions/runs/9674533040/job/26696570584?pr=27593), it looks like the test suite runs all the post-migrations as well. I'm going remove the 'drop db column' post-migration from this PR. If plugins need to time to update, then it makes more sense to not keep a destructive commit available for now.

I'll also see if I can get a cleaner full test suite run locally to reduce the back-and-forth of waiting for CI to run.

counterleft commented:
Alright, a lower number of test failures [this time around](https://github.com/discourse/discourse/actions/runs/9687901420?pr=27593). Looks like the problems are:

1. I missed a i18n default setting for allowing "hide presence".
2. The model schema annotation was wrong.
3. I missed a chat spec that should have used `hide_presence` instead of `hide_profile`.
4. A flakey test for private_messages_map would work sometimes depending on whether the user `<div>` rendered in a certain order.

I have fixes incoming.

<img width="614" alt="Screenshot 2024-06-27 at 17 54 01" src="https://github.com/discourse/discourse/assets/4782/f8d6368e-b702-4f21-9a53-b89c2fc536db">

<img width="676" alt="Screenshot 2024-06-27 at 12 11 20" src="https://github.com/discourse/discourse/assets/4782/05011a27-3b93-492d-990f-32fc1332c4a8">

counterleft commented:
The Check Annotations check isn't happy with me. Going to apply the diff it suggests directly and re-push.

counterleft commented:
Given this [update on the Community Discourse](https://meta.discourse.org/t/split-profile-hiding-and-presence/184316/21?u=counterleft), it sounds like the core team is going to work on this.

I'm going to close this PR because it seems like it won't be receiving any more support. Cheers, folks.
